### ISSUE TITLE
mypy tests: Allow per distribution strictness options

### ISSUE THREAD

**Issue by JelleZijlstra** [python/typeshed#1526]
We should try turning on the `--disallow-incomplete-defs` flag from python/mypy#3744. This can catch issues where we forget type annotations for some function arguments in a stub.

**Comment by gvanrossum** [comment]
Good idea.

**Comment by ilinum** [comment]
I agree :) 

There are a lot of errors (little over 5000). 
[Here](https://gist.github.com/ilinum/c5fcbf184ae3573c7646a7b544041461) is the gist with all the errors it produces.

**Comment by OddBloke** [comment]
Considering only the `mypy --python-version 3.6 --strict-optional --disallow-incomplete-defs` failures, the breakdown between stdlib/third-party is

```sh
$ cut -d/ -f1 < out | sort | uniq -c
     77 stdlib
    895 third_party
```

so enabling this just for the stdlib initially might make addressing them more tractable?

(Full breakdown by module at https://gist.github.com/OddBloke/dbd78409dcd53bdbb6b3b8571bd29720)

**Comment by CraftSpider** [comment]
I'm interested in tackling at least the stdlib side of this. Would it be better to do this as individual, more well thought-out PRs for each module, or one much larger commit that only does relatively general types

**Comment by srittau** [comment]
In general, smaller PRs are easier to review, but too many PRs can also become cumbersome. It might be best to have one PR per package if a package has a few changes needed and a few PRs for packages that only have few changes.

**Comment by hauntsaninja** [comment]
Note `--disallow-untyped-defs` is actually stricter than `--disallow-incomplete-defs` and should be preferred.

**Comment by srittau** [comment]
I am not sure we should enable either. I prefer to have unannotated types over using `Any`. I also prefer unannotated types over not having definitions or using "incomplete" markers. Enabling these warnings would raise the bar for contributions, with more complex libraries significantly.

**Comment by hauntsaninja** [comment]
I agree on both points! (Although it's reasonable to aspire to turning this on as a lint for stdlib one day, since we're close to completion and hopefully the stdlib isn't changing too drastically).

My point with the above was just that `--disallow-untyped-defs` is more in line with what I want for identifying stubs that need improvement (in the stub context, it can be a surprise that `--disallow-incomplete-defs` doesn't complain about definitions that are missing types entirely).

**Comment by srittau** [comment]
Good point about stdlib.

**Comment by JukkaL** [comment]
Once we've migrated to modular typeshed, it should be easy to add support for specifying  stricness options for each distribution separately in the metadata file. This wouldn't directly help with the stdlib, since stdlib would be distributed as a single entity.

**Comment by srittau** [comment]
I renamed this ticket to what I believe the consensus of the discussion is. Please revert if you don't agree.

**Comment by srittau** [comment]
I've removed the "deferred" label as this could now be implemented.

**Comment by JelleZijlstra** [comment]
Fixed in #5169
